### PR TITLE
Migrate from Gson to Jackson

### PR DESCRIPTION
**Pull request by alexis779**
Currently JSON serialization relies on *Gson* library. This change enables seamless serialization through *Jackson* library, which is also widely used, for example in **Spring Boot** framework.

Refer to the newly added unit test. It validates the Java object to JSON string and vice versa transformations of 2 main models,
- ChatCompletionRequest
- ChatCompletionResponse

For each models, the instance creation process relies on the builder design pattern. Jackson can be configured to use it by adding following annotation at the class level 
```
@JsonDeserialize(builder = ChatCompletionRequest.Builder.class)
```

Every builder class also needs following annotations:
```
@JsonPOJOBuilder(withPrefix = "")
@JsonIgnoreProperties(ignoreUnknown = true)

```

Because of this custom deserializer, Jackson requires adding an annotation at the field level, for each fields that need to be included in the JSON output,

```
    @JsonProperty
    private final String model;
```

Following annotation avoids outputting null fields:
```
@JsonInclude(JsonInclude.Include.NON_NULL)
```

Following annotation serializes camel case to snake case. It also deserializes json snake case attributes back to camel case Java fields.
```
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
```

For polymorphism, following annotations at the interface level specifies which field to lookup to identify type. They also register each available implementations.
```
@JsonTypeInfo(
        use = JsonTypeInfo.Id.NAME,
        property = "role")
@JsonSubTypes({
        @JsonSubTypes.Type(value = AssistantMessage.class, name = "assistant"),
        @JsonSubTypes.Type(value = SystemMessage.class, name = "system"),
        @JsonSubTypes.Type(value = ToolMessage.class, name = "tool"),
        @JsonSubTypes.Type(value = UserMessage.class, name = "user")
})
```

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -204,6 +210,14 @@
                     </execution>
                 </executions>
             </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <configuration>
+                    <source>9</source>
```
Comment (langchain4j) [inline pom.xml:217]: Why is this required? We are still java8-compatible

File: src/main/java/dev/ai4j/openai4j/chat/AssistantMessage.java [hunk 2]
```diff
@@ -1,19 +1,32 @@
...
+
 import java.util.List;
 import java.util.Objects;
 
 import static dev.ai4j.openai4j.chat.Role.ASSISTANT;
 import static java.util.Arrays.asList;
 import static java.util.Collections.unmodifiableList;
 
+@JsonDeserialize(builder = AssistantMessage.Builder.class)
+@JsonInclude(JsonInclude.Include.NON_NULL)
```
Comment (langchain4j) [inline src/main/java/dev/ai4j/openai4j/chat/AssistantMessage.java:17]: We also need to add `@JsonIgnoreProperties(ignoreUnknown = true)` everywhere to make sure deserialization will not fail when OpenAI will add some new fields.

**Review by langchain4j — COMMENTED** [review]
@alexis779 thank you!

Do you plan to finish the migration? It would be great to get rid of Gson completely and start using Jackson everywhere. Right now Gson is still used...

**Comment by langchain4j** [comment]
Please also make sure all fields are serialized with snake case strategy

File: src/main/java/dev/ai4j/openai4j/chat/AssistantMessage.java (hunk 2 above)
Comment (alexis779) [reply src/main/java/dev/ai4j/openai4j/chat/AssistantMessage.java:21]: Added the annotation to accept unknown fields when deserializing JSON into the Java model.

**Comment by alexis779** [comment]
> Please also make sure all fields are serialized with snake case strategy

Added following annotation at the class level and on the builder, for both serialization and deserialization.
```
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
```

**Comment by alexis779** [comment]
> @alexis779 thank you!
> 
> Do you plan to finish the migration? It would be great to get rid of Gson completely and start using Jackson everywhere. Right now Gson is still used...

Removed Gson dependency to move over to Jackson exclusively.

**Review by langchain4j — COMMENTED** [review]
@alexis779 thanks a lot, great job!

I've noticed that `ImageDetail` and `ToolChoiceMode` were not covered, could you please check?

Also, requests are now not pretty-printed, could you please revert it back?

Thanks!

File: src/main/java/dev/ai4j/openai4j/chat/AssistantMessage.java [hunk 3]
```diff
@@ -122,6 +143,7 @@ public Builder name(String name) {
             return this;
         }
 
+        @JsonSetter
```
[the remark this answers is not in the document]
Comment (alexis779) [reply src/main/java/dev/ai4j/openai4j/chat/AssistantMessage.java]: Moved the annotation from the varargs to the List parameter.

File: src/main/java/dev/ai4j/openai4j/chat/AssistantMessage.java [hunk 4]
```diff
@@ -100,6 +118,9 @@ public static Builder builder() {
         return new Builder();
     }
 
+    @JsonPOJOBuilder(withPrefix = "")
+    @JsonIgnoreProperties(ignoreUnknown = true)
+    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
```
[the remark this answers is not in the document]
Comment (alexis779) [reply src/main/java/dev/ai4j/openai4j/chat/AssistantMessage.java:124]: Yes, it's required on both.

File: src/main/java/dev/ai4j/openai4j/chat/ChatCompletionRequest.java [hunk 5]
```diff
@@ -294,6 +329,7 @@ public Builder messages(List<Message> messages) {
             return this;
         }
 
+        @JsonSetter
```
[the remark this answers is not in the document]
Comment (alexis779) [reply src/main/java/dev/ai4j/openai4j/chat/ChatCompletionRequest.java]: done

File: src/main/java/dev/ai4j/openai4j/chat/JsonSchemaElement.java [hunk 6]
```diff
@@ -1,5 +1,25 @@
...
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonSubTypes;
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import com.fasterxml.jackson.databind.PropertyNamingStrategies;
+import com.fasterxml.jackson.databind.annotation.JsonNaming;
+
+@JsonTypeInfo(
+        use = JsonTypeInfo.Id.NAME,
+        property = "type")
+@JsonSubTypes({
```
[the remark this answers is not in the document]
Comment (alexis779) [reply src/main/java/dev/ai4j/openai4j/chat/JsonSchemaElement.java:13]: You're right this is used for deserialization only.

I would recommend keeping it as we want to support both ways, at least for testing purposes.

File: src/main/java/dev/ai4j/openai4j/chat/Message.java [hunk 7]
```diff
@@ -1,5 +1,23 @@
...
+import com.fasterxml.jackson.annotation.JsonSubTypes;
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import com.fasterxml.jackson.databind.PropertyNamingStrategies;
+import com.fasterxml.jackson.databind.annotation.JsonNaming;
+
+@JsonTypeInfo(
+        use = JsonTypeInfo.Id.NAME,
+        property = "role")
+@JsonSubTypes({
+        @JsonSubTypes.Type(value = AssistantMessage.class, name = "assistant"),
```
[the remark this answers is not in the document]
Comment (alexis779) [reply src/main/java/dev/ai4j/openai4j/chat/Message.java:13]: - same comment: we need it for deserialization only for testing purpose mostly ...

- added annotation on role field as well.

**Review by langchain4j — APPROVED** [review]
@alexis779 thanks a lot, great job!
